main.tf

```diff
@@ -1,98 +1,72 @@
-  namespace_context_or_default           = "${length(local.namespaces[0]) > 0 ? local.namespaces[0] : var.namespace}"
-  namespace_or_context                   = "${var.namespace != "" ? var.namespace : local.namespace_context_or_default}"
-  namespace                              = "${lower(replace(local.namespace_or_context, local.regex_replace_chars, ""))}"
-  environments                           = "${concat(local.context_local["environment"], list(""))}"
-  environment_context_or_default         = "${length(local.environments[0]) > 0 ? local.environments[0] : var.environment}"
-  environment_or_context                 = "${var.environment != "" ? var.environment : local.environment_context_or_default}"
-  environment                            = "${lower(replace(local.environment_or_context, local.regex_replace_chars, ""))}"
-  stages                                 = "${concat(local.context_local["stage"], list(""))}"
-  stage_context_or_default               = "${length(local.stages[0]) > 0 ? local.stages[0] : var.stage}"
-  stage_or_context                       = "${var.stage != "" ? var.stage : local.stage_context_or_default}"
-  stage                                  = "${lower(replace(local.stage_or_context, local.regex_replace_chars, ""))}"
-  delimiters                             = "${concat(local.context_local["delimiter"], list(""))}"
-  delimiter_context_or_default           = "${length(local.delimiters[0]) > 0 ? local.delimiters[0] : var.delimiter}"
-  delimiter                              = "${var.delimiter != "-" ? var.delimiter : local.delimiter_context_or_default}"
   # Merge attributes
-  attributes = ["${distinct(compact(concat(var.attributes, local.context_local["attributes"])))}"]
-  # Generate tags (don't include tags with empty values)
-  generated_tags = "${zipmap(
-    compact(list("Name", local.namespace != "" ? "Namespace" : "", local.environment != "" ? "Environment" : "", local.stage != "" ? "Stage" : "")),
-    compact(list(local.id, local.namespace, local.environment, local.stage))
-    )}"
-  tags                     = "${merge(zipmap(local.context_local["tags_keys"], local.context_local["tags_values"]), local.generated_tags, var.tags)}"
-  tags_as_list_of_maps     = ["${data.null_data_source.tags_as_list_of_maps.*.outputs}"]
-  label_order_default_list = "${list("namespace", "environment", "stage", "name", "attributes")}"
-  label_order_context_list = "${distinct(compact(local.context_local["label_order"]))}"
-  label_order_final_list   = ["${distinct(compact(coalescelist(var.label_order, local.label_order_context_list, local.label_order_default_list)))}"]
-  label_order_length       = "${(length(local.label_order_final_list))}"
+  attributes = compact(distinct(concat(var.attributes, var.context.attributes, local.defaults.attributes)))
+
+  generated_tags = { for l in keys(local.id_context) : title(l) => local.id_context[l] if length(local.id_context[l]) > 0 }
```

Comment:
wow, TF 0.12 shines here :)

Reply:
Yes, I liked this. Reminded me of the `map { ... }` expression in ruby and perl.